src/cid_collections/hash_map.rs

```diff
@@ -0,0 +1,321 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::{CidV1DagCborBlake2b256, MaybeCompactedCid, Uncompactable};
+use cid::Cid;
+use std::collections::hash_map::{
+    Entry as StdEntry, IntoIter as StdIntoIter, OccupiedEntry as StdOccupiedEntry,
+    VacantEntry as StdVacantEntry,
+};
+
+/// A space-optimised hashmap of [`Cid`]s.
+///
+/// We accept the implementation complexity of per-compaction-method HashMaps for
+/// the space savings.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CidHashMap<V> {
+    compact: ahash::HashMap<CidV1DagCborBlake2b256, V>,
+    uncompact: ahash::HashMap<Uncompactable, V>,
+}
+
+impl<V> CidHashMap<V> {
```

Comment:
All of these methods require documentation.

Reply:
I would document these by forwarding to the standard library (I don't want to copy-paste the docs) - is that definitely what you want?